packages/native/src/useLinking.tsx

```diff
@@ -116,203 +306,143 @@ export default function useLinking(
-        } else {
-          return;
-        }
-      }
+      previousStateRef.current = state;
+      pendingPopStatePathRef.current = undefined;
 
-      previousStateLengthRef.current = stateLength;
+      // To detect the kind of state change, we need to:
+      // - Find the common focused navigation state in previous and current state
+      // - If only the route keys changed, compare history/routes.length to check if we go back/forward/replace
+      // - If no common focused navigation state found, it's a replace
+      const [previousFocusedState, focusedState] = findMatchingState(
+        previousState,
+        state
+      );
 
       if (
-        pendingStateUpdateRef.current &&
-        location.pathname + location.search === path
+        previousFocusedState &&
+        focusedState &&
+        // We should only handle push/pop if path changed from what was in last `popstate`
+        // Otherwise it's likely a change triggered by `popstate`
+        path !== pendingPath
       ) {
-        pendingStateUpdateRef.current = false;
-        return;
-      }
-
-      let index = history.state?.index ?? 0;
-
-      if (previousStateLength === stateLength) {
-        // If no new entries were added to history in our navigation state, we want to replaceState
-        if (location.pathname + location.search !== path) {
-          history.replaceState({ index }, '', path);
-          previousHistoryIndexRef.current = index;
-        }
-      } else if (stateLength > previousStateLength) {
-        // If new entries were added, pushState until we have same length
-        // This won't be accurate if multiple entries were added at once, but that's the best we can do
-        for (let i = 0, l = stateLength - previousStateLength; i < l; i++) {
-          index++;
-          history.pushState({ index }, '', path);
-        }
+        const historyDelta =
+          (focusedState.history
+            ? focusedState.history.length
+            : focusedState.routes.length) -
+          (previousFocusedState.history
+            ? previousFocusedState.history.length
+            : previousFocusedState.routes.length);
+
+        if (historyDelta > 0) {
+          // If history length is increased, we should pushState
+          // Note that path might not actually change here, for example, drawer open should pushState
+          history.push({ path, state });
+        } else if (historyDelta < 0) {
+          // If history length is decreased, i.e. entries were removed, we want to go back
+
+          const nextIndex = history.backIndex({ path });
+          const currentIndex = history.index;
+
+          if (nextIndex !== -1 && nextIndex < currentIndex) {
+            // An existing entry for this path exists and it's less than current index, go back to that
+            await history.go(nextIndex - currentIndex);
+          } else {
+            // We couldn't find an existing entry to go back to, so we'll go back by the delta
```

Comment:
can you explain why it can happen? Maybe we can just `go` in this case?

Reply:
The user could change the screen configuration or linking configuration which would result in different path, or user has specified some initial screens with existing history so we haven't actually visited this `path` etc. I think going by delta would be more correct here.

Reply:
Hey, sorry to jump in here completely out of the blue, but I've been following this PR since the very beginning because I was actually tracking down a bug I experienced when using linking with nested navigators on web that originates exactly around this part of the code. You guys might already know about it, but I figured there's no harm on pointing it out, so here it is.

Lets say you have a nested navigation tree that looks sort of like this (where N=navigator and S=screen):

N1
|- N2
|     |- S3
|     |- S4
|     |- S5
|     |- S6
| 
|- S1
|- S2

If, by using linking, you directly access any screen on N2, let's say S3 directly by url. Then you navigate through a couple of screens on the same level, let's say first to S4 and then to S6 (at which point your history length becomes at least 2). If you then try to navigate to S1, this piece of the code was actually comparing the history length (which could potentially be any number) with the length of the newly created state as returned from "getStateFromPath" for S1 via "getStateLength". Since the state for S1 was new, it had no history, so its length was always 1 (the depth of its nesting). Therefore, your previous history length was always greater than your current state length, causing this to always trigger a "history.go(-delta);" which made the url point to a previously visited url (S4 following my example), instead of S1's url as it should.

Since the navigation itself is independent of this piece of core, the site continues to work properly, except for the fat that urls from that point onwards are completely miss-matched with the screens being rendered.

Sorry for the long message but I wanted to make it as clear as possible. If you where already aware of this issue, feel free to completely disregard this message! Or if you need more info on how to reproduce it or whatever, don't hesitate to ask me. It caused me more than one headache last week, and I was very happy to see work being done on this file.

Thanks!

Reply:
@martindido sounds like the linked issues